batch_finder_resource_method.md

```diff
@@ -0,0 +1,418 @@
+-   Arrays of one of the types above, e.g. `String[]`, `long[]`, ...
+
+Valid type for batch criteria parameter:
+
+- Can only be Arrays of Record template type, if have to use some other data types like Pegasus Enum, etc as the array item,
+ need to wrap it into a Record Template (`.pdsc` schema)
+ 
+### Criteria Filter
+To implement a batch finder, the resource owner has to define a `RecordTemplate` to define a criteria filter parameter.
+The batch finder method will have to accept a array of this criteria filter.
+
+Example:
+
+The "GreetingCriteria" class represent a criteria filter to filter by "id" or by "tone".
+This class is auto-generated from the pdsc. 
+```java
+public class GreetingCriteria extends RecordTemplate
+{
+
+    private final static GreetingCriteria.Fields _fields = new GreetingCriteria.Fields();
+    private final static RecordDataSchema SCHEMA = ((RecordDataSchema) DataTemplateUtil.parseSchema("{\"type\":\"record\",\"name\":\"GreetingCriteria\",\"namespace\":\"com.linkedin.restli.examples.greetings.api\",\"doc\":\"A search criteria to filter greetings.\",\"fields\":[{\"name\":\"id\",\"type\":\"long\",\"doc\":\"Greeting ID to filter on\"},{\"name\":\"tone\",\"type\":{\"type\":\"enum\",\"name\":\"Tone\",\"symbols\":[\"FRIENDLY\",\"SINCERE\",\"INSULTING\"]},\"doc\":\"Greeting tone to filter on\"}]}"));
+    private final static RecordDataSchema.Field FIELD_Id = SCHEMA.getField("id");
+    private final static RecordDataSchema.Field FIELD_Tone = SCHEMA.getField("tone");
+    ...
+}
+```
+
+The resource owner need to define their own search criteria `.pdsc` file.
+```
+{
+ "type" : "record",
+ "name" : "GreetingCriteria",
+ "namespace" : "com.linkedin.restli.examples.greetings.api",
+ "doc" : "A search criteria to filter greetings.",
+ "fields" : [
+   {
+     "name" : "id",
+     "doc": "Greeting ID to filter on",
+     "type" : "long"
+   },
+   {
+     "name" : "tone",
+     "doc" : "Greeting tone to filter on",
+     "type" : "Tone"
+   }
+ ]
+}
+```
+ 
+### BatchFinderResult
+BATCH_Finder methods must return `BatchFinderResult<QK extends RecordTemplate, V extends RecordTemplate, MD extends RecordTemplate>`:
+
+- `QK` : The type of the BATCH_FINDER criteria filter
+- `V` :  The type of the resource, aka, the entity type
+- `MD` : The type of the meta data, if do not need metadata, just set it `EmptyRecord`
+
+For each search criteria in the BatchFinderRequest, it can get either a successful reponse
+which is a `CollectionResult`(a list of entities), Or an error/failure which maybe represented by
+a `RestLiServiceException`, which will be wrapped into an `ErrorResponse` later when building BatchFinderResponse
+to return to client.
+```
+public class BatchFinderResult<QK,V extends RecordTemplate,MD extends RecordTemplate>
+{
+   private final Map<QK,CollectionResult<V,R>> _elements;
+   private final Map<QK,RestLiServiceException> _errors;
+   ...
+}
+```
+
+## Future consideration
+### Support primitive type(not ready)
```

Comment:
I would say not supported. For now, we don't want to support this feature that will prevent any evolutions.

Reply:
Got it.